### PR TITLE
bugfix- updated API to fetch resources based on resource name

### PR DESCRIPTION
**Pull request by SINIKI**
## Description
Previously, we were fetching the complete list of resource groups that user had access to. In this PR, I have updated the API to filter resources based on name and fetch only relevant resource details.

## Checklist
- [ ] I have read the instructions mentioned in [Contribute to Code](/Contributing.md).
- [ ] I have read and understood the criteria described under [submitting changes](/Contributing.md#submitting-changes). 
- [ ] The title of the PR clearly describes the intent of the PR.
- [ ] This PR does not introduce any breaking changes to the code.

### CODE REVIEW

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 [hunk 1]
```diff
@@ -109,76 +109,55 @@ class PIM: CommandBase {
...
+        # Note: At present, we do not provide PIM operation management for management group. However, if needed in the future, it can be added in the else statement. >> $filter=(type%20eq%20%27managementgroup%27)
+        
+        if($type -eq 'resource') {
+            $resourceUrl = $this.APIroot + "/resources?`$select=id,displayName,type&`$filter=(type%20ne%20%27resourcegroup%27%20and%20type%20ne%20%27subscription%27%20and%20type%20ne%20%27managementgroup%27)%20and%20contains(tolower(displayName),%20%27{0}%27)" -f $resourceName.ToLower()
+        }
+        elseif($type -eq 'subscription')
+        {
+            $resourceUrl = $this.APIroot + "/resources?`$select=id,displayName,type&`$filter=(type%20eq%20%27subscription%27)&`$orderby=type"
+        }
+        elseif($type -eq 'resourcegroup')
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/PIM/PIMScript.ps1]: Code comments are missing. You should explain about each block and the purpose of the API etc.

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 [hunk 2]
```diff
@@ -109,76 +109,55 @@ class PIM: CommandBase {
...
+        {
+            $resourceUrl = $this.APIroot + "/resources?`$select=id,displayName,type&`$filter=(type%20eq%20%27resourcegroup%27)%20and%20contains(tolower(displayName),%20%27{0}%27)&`$orderby=type" -f $resourceName.ToLower()
+        }
+        else
+        {
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/PIM/PIMScript.ps1]: What is purpose of else block here

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 [hunk 3]
```diff
@@ -219,7 +198,7 @@ class PIM: CommandBase {
     #List Assignment
     hidden [PSObject] ListAssignmentsWithFilter($resourceId, $IsPermanent) {
         $this.AcquireToken()
-        $url = $this.APIroot + "resources/" + $resourceId + "`/roleAssignments?`$expand=subject,roleDefinition(`$expand=resource)"
+        $url = $this.APIroot + "/resources/" + $resourceId + "`/roleAssignments?`$expand=subject,roleDefinition(`$expand=resource)"
         #Write-Host $url
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/PIM/PIMScript.ps1]: Remove commented code

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 [hunk 4]
```diff
@@ -219,7 +198,7 @@ class PIM: CommandBase {
...
 
         $response = Invoke-WebRequest -UseBasicParsing -Headers $this.headerParams -Uri $url -Method Get
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/PIM/PIMScript.ps1]: Use WebRequest Helper

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 [hunk 5]
```diff
@@ -198,7 +177,7 @@ class PIM: CommandBase {
     #List roles
     hidden [PSObject] ListRoles($resourceId) {
         $this.AcquireToken();
-        $url = $this.APIroot + "resources/" + $resourceId + "/roleDefinitions?`$select=id,displayName,type,templateId,resourceId,externalId,subjectCount,eligibleAssignmentCount,activeAssignmentCount&`$orderby=activeAssignmentCount%20desc"
+        $url = $this.APIroot + "/resources/" + $resourceId + "/roleDefinitions?`$select=id,displayName,type,templateId,resourceId,externalId,subjectCount,eligibleAssignmentCount,activeAssignmentCount&`$orderby=activeAssignmentCount%20desc"
         $response = Invoke-WebRequest -UseBasicParsing -Headers $this.headerParams -Uri $url -Method Get
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/PIM/PIMScript.ps1]: Use WebRequest Helper

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 [hunk 6]
```diff
@@ -109,76 +109,55 @@ class PIM: CommandBase {
...
+        $response = $null
+        try
+        {
+            $response = Invoke-WebRequest -UseBasicParsing -Headers $this.headerParams -Uri $resourceUrl -Method Get
+            $values = ConvertFrom-Json $response.Content
+            $resources = $values.value
+            $hasOdata = $values | Get-Member -Name '@odata.nextLink'
+            while ($null -ne $hasOdata -and -not([string]::IsNullOrEmpty(($values).'@odata.nextLink')))
+            {
+                $response = Invoke-WebRequest -UseBasicParsing -Headers $this.headerParams -Uri $(($values).'@odata.nextLink') -Method Get
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/PIM/PIMScript.ps1]: Use WebRequest Helper

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 (hunk 2 above)
Comment (SINIKI) [reply src/AzSK/Framework/Core/PIM/PIMScript.ps1]: Removed else block

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 (hunk 1 above)
Comment (SINIKI) [reply src/AzSK/Framework/Core/PIM/PIMScript.ps1]: Added comments

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 (hunk 3 above)
Comment (SINIKI) [reply src/AzSK/Framework/Core/PIM/PIMScript.ps1]: Deleted comment

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 (hunk 6 above)
Comment (SINIKI) [reply src/AzSK/Framework/Core/PIM/PIMScript.ps1]: I will raise separate PR for this.

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 (hunk 5 above)
Comment (SINIKI) [reply src/AzSK/Framework/Core/PIM/PIMScript.ps1]: I will raise separate PR for this.

File: src/AzSK/Framework/Core/PIM/PIMScript.ps1 (hunk 4 above)
Comment (SINIKI) [reply src/AzSK/Framework/Core/PIM/PIMScript.ps1]: I will raise separate PR for this.

**Pull request merged**